os/drivers/usbhost/uvc/usbhost_uvc_video.c

```diff
@@ -0,0 +1,1190 @@
+/****************************************************************************
+ *
+ * Copyright 2019 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+* Included Files
+****************************************************************************/
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <semaphore.h>
+#include <assert.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/usb/usb.h>
+#include <tinyara/usb/usbhost.h>
+#include "usbhost_uvc.h"
+
+static void usbhost_transfer_callback(FAR void *arg, ssize_t nbytes);
+FAR struct usbhost_uvc_state_s *g_usbuvc_handoff;
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+/****************************************************************************
+* Name: uvc_circ_bbuf_init
+*
+* Description:
+*   A helper funtion to initialize circular queue.
+*
+* Input Parameters:
+*   c - pointer to circular queue object.
+*
+* Returned Values:
+*   0 on success, -1 if queue is full.
+*
+****************************************************************************/
+void uvc_circ_bbuf_init(struct uvc_circ_buf *c)
+{
+	int i = 0;
+	c->head = -1;
+	c->tail = -1;
+	c->maxlen = UVC_MAX_TRF_BUF;
+	while (i < c->maxlen) {
+		c->bobj[i] = NULL;
+		i++;
+	}
+}
+
+/****************************************************************************
+ * Name: uvc_circ_bbuf_push
+ *
+ * Description:
+ *   A helper funtion to push the buffer object to circular queue.
+ *
+ * Input Parameters:
+ *   c - pointer to circular queue object.
+ *   buf - pointer to buffer object.
+ *
+ * Returned Values:
+ *   0 on success, -1 if queue is full.
+ *
+ ****************************************************************************/
+static inline int uvc_circ_bbuf_push(struct uvc_circ_buf *c, struct uvc_buf_object *buf)
+{
+	if ((c->head == 0 && c->tail == c->maxlen - 1) || (c->tail == (c->head - 1) % (c->maxlen - 1))) {
```

Comment:
Is it okay not to check the validation of `c`? It can be dereferencing.

Reply:
It cant be dereferenced, as we pass the address of fbuf or cbuf from struct uvc_streaming object. Hence there is no need to check here in this function.